core/src/main/java/org/axonframework/messaging/annotation/AnnotatedHandlerInspector.java

```diff
@@ -68,64 +71,85 @@ private AnnotatedHandlerInspector(Class<T> inspectedType,
+                                                              handlerDefinition,
+                                                              registry));
         }
         //noinspection unchecked
         return registry.get(inspectedType);
     }
 
     private static <T> AnnotatedHandlerInspector<T> initialize(Class<T> inspectedType,
                                                                ParameterResolverFactory parameterResolverFactory,
+                                                               HandlerDefinition handlerDefinition,
                                                                Map<Class<?>, AnnotatedHandlerInspector> registry) {
         List<AnnotatedHandlerInspector<? super T>> parents = new ArrayList<>();
         for (Class<?> iFace : inspectedType.getInterfaces()) {
             //noinspection unchecked
-            parents.add(createInspector(iFace, parameterResolverFactory, registry));
+            parents.add(createInspector(iFace,
+                                        parameterResolverFactory,
+                                        handlerDefinition,
+                                        registry));
         }
         if (inspectedType.getSuperclass() != null && !Object.class.equals(inspectedType.getSuperclass())) {
-            parents.add(createInspector(inspectedType.getSuperclass(), parameterResolverFactory, registry));
+            parents.add(createInspector(inspectedType.getSuperclass(),
+                                        parameterResolverFactory,
+                                        handlerDefinition,
+                                        registry));
         }
-        AnnotatedHandlerInspector<T> inspector =
-                new AnnotatedHandlerInspector<>(inspectedType, parents, parameterResolverFactory, registry);
-        inspector.initializeMessageHandlers(parameterResolverFactory);
+        AnnotatedHandlerInspector<T> inspector = new AnnotatedHandlerInspector<>(inspectedType,
+                                                                                 parents,
+                                                                                 parameterResolverFactory,
+                                                                                 handlerDefinition,
+                                                                                 registry);
+        inspector.initializeMessageHandlers(parameterResolverFactory, handlerDefinition);
         return inspector;
     }
 
-    private void initializeMessageHandlers(ParameterResolverFactory parameterResolverFactory) {
-        List<HandlerDefinition> definitions = new ArrayList<>();
-        ServiceLoader.load(HandlerDefinition.class).forEach(definitions::add);
-        List<HandlerEnhancerDefinition> wrapperDefinitions = new ArrayList<>();
-        ServiceLoader.load(HandlerEnhancerDefinition.class).forEach(wrapperDefinitions::add);
```

Comment:
Why is the `List<HandlerEnhancerDefinition>` removed? Because this `AnnotatedHandlerInspector` will get a `MultiHandlerDefinition` which contains the `HandlerEnhancerDefinition` in turn, right?

Reply:
Yup, my approach was with list of definitions, @snowe2010 had a different approach which I liked better.